File: src/aws-tools.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js"
import { awsRegionArgs } from "./args.js"
import {
  EksAddonInput,
  EksClusterInput,
  EksNodegroupInput,
  RegionSchema
} from "./schemas.js"
import { commandTool } from "./tool-runner.js"
import type { RuntimeConfig } from "./types.js"

export function registerAwsTools(server: McpServer, config: RuntimeConfig): void {
  server.registerTool(
    "aws_eks_list_clusters",
    {
      description: "List EKS clusters in an explicit AWS region.",
      inputSchema: { region: RegionSchema }
    },
    ({ region }) => commandTool(config, { command: "aws", args: ["eks", "list-clusters", ...awsRegionArgs(region)] })
  )

  server.registerTool(
    "aws_eks_describe_cluster",
    {
      description: "Describe one EKS cluster in an explicit region.",
      inputSchema: EksClusterInput.shape
    },
    ({ region, cluster }) =>
      commandTool(config, {
        command: "aws",
        args: ["eks", "describe-cluster", "--name", cluster, ...awsRegionArgs(region)]
      })
  )

  server.registerTool(
    "aws_eks_list_nodegroups",
    {
      description: "List node groups for one EKS cluster.",
      inputSchema: EksClusterInput.shape
    },
    ({ region, cluster }) =>
      commandTool(config, {
        command: "aws",
        args: ["eks", "list-nodegroups", "--cluster-name", cluster, ...awsRegionArgs(region)]
      })
  )

  server.registerTool(
    "aws_eks_describe_nodegroup",
    {
      description: "Describe one EKS node group.",
      inputSchema: EksNodegroupInput.shape
    },
    ({ region, cluster, nodegroup }) =>
      commandTool(config, {
        command: "aws",
        args: [
          "eks",
          "describe-nodegroup",
          "--cluster-name",
          cluster,
          "--nodegroup-name",
          nodegroup,
          ...awsRegionArgs(region)
        ]
      })
  )

  server.registerTool(
    "aws_eks_list_addons",
    {
      description: "List add-ons installed on one EKS cluster.",
      inputSchema: EksClusterInput.shape
    },
    ({ region, cluster }) =>
      commandTool(config, {
        command: "aws",
        args: ["eks", "list-addons", "--cluster-name", cluster, ...awsRegionArgs(region)]
      })
  )

  server.registerTool(
    "aws_eks_describe_addon",
    {
      description: "Describe one EKS add-on.",
      inputSchema: EksAddonInput.shape
    },
    ({ region, cluster, addon }) =>
      commandTool(config, {
        command: "aws",
        args: [
          "eks",
          "describe-addon",
          "--cluster-name",
          cluster,
          "--addon-name",
          addon,
          ...awsRegionArgs(region)
        ]
      })
  )

  server.registerTool(
    "eks_generate_kubeconfig_dry_run",
    {
      description: "Generate kubeconfig content with aws eks update-kubeconfig --dry-run only.",
      inputSchema: EksClusterInput.shape
    },
    ({ region, cluster }) =>
      commandTool(config, {
        command: "aws",
        args: ["eks", "update-kubeconfig", "--name", cluster, "--dry-run", ...awsRegionArgs(region)]
      })
  )
}

File: safe_eks_mcp/aws_tools.py
from __future__ import annotations

from .args import aws_region_args
from .runner import command_tool
from .types import CliInvocation, RuntimeConfig, ToolPayloadJson
from .validation import require_cluster, require_name, require_region


def aws_eks_list_clusters(config: RuntimeConfig, region: str) -> ToolPayloadJson:
    parsed_region = require_region(region)
    return command_tool(
        config, CliInvocation(command="aws", args=("eks", "list-clusters", *aws_region_args(parsed_region)))
    )


def aws_eks_describe_cluster(config: RuntimeConfig, region: str, cluster: str) -> ToolPayloadJson:
    return command_tool(
        config,
        CliInvocation(
            command="aws",
            args=(
                "eks",
                "describe-cluster",
                "--name",
                require_cluster(cluster),
                *aws_region_args(require_region(region)),
            ),
        ),
    )


def aws_eks_list_nodegroups(config: RuntimeConfig, region: str, cluster: str) -> ToolPayloadJson:
    return command_tool(
        config,
        CliInvocation(
            command="aws",
            args=(
                "eks",
                "list-nodegroups",
                "--cluster-name",
                require_cluster(cluster),
                *aws_region_args(require_region(region)),
            ),
        ),
    )


def aws_eks_describe_nodegroup(
    config: RuntimeConfig, region: str, cluster: str, nodegroup: str
) -> ToolPayloadJson:
    return command_tool(
        config,
        CliInvocation(
            command="aws",
            args=(
                "eks",
                "describe-nodegroup",
                "--cluster-name",
                require_cluster(cluster),
                "--nodegroup-name",
                require_name(nodegroup, "nodegroup"),
                *aws_region_args(require_region(region)),
            ),
        ),
    )


def aws_eks_list_addons(config: RuntimeConfig, region: str, cluster: str) -> ToolPayloadJson:
    return command_tool(
        config,
        CliInvocation(
            command="aws",
            args=(
                "eks",
                "list-addons",
                "--cluster-name",
                require_cluster(cluster),
                *aws_region_args(require_region(region)),
            ),
        ),
    )


def aws_eks_describe_addon(config: RuntimeConfig, region: str, cluster: str, addon: str) -> ToolPayloadJson:
    return command_tool(
        config,
        CliInvocation(
            command="aws",
            args=(
                "eks",
                "describe-addon",
                "--cluster-name",
                require_cluster(cluster),
                "--addon-name",
                require_name(addon, "addon"),
                *aws_region_args(require_region(region)),
            ),
        ),
    )


def eks_generate_kubeconfig_dry_run(config: RuntimeConfig, region: str, cluster: str) -> ToolPayloadJson:
    return command_tool(
        config,
        CliInvocation(
            command="aws",
            args=(
                "eks",
                "update-kubeconfig",
                "--name",
                require_cluster(cluster),
                "--dry-run",
                *aws_region_args(require_region(region)),
            ),
        ),
    )
